revolution_pi/gpio_pin.go

```diff
@@ -11,37 +11,83 @@ import (
+	Length       uint16 // length of the variable in bits. Possible values are 1, 8, 16 and 32
+	ControlChip  *gpioChip
+	pwmMode      bool
+	initialized  bool
+	outputOffset uint16
+	inputOffset  uint16
 }
 
 func (pin *gpioPin) initialize() error {
-	val, err := pin.ControlChip.getBitValue(int64(110+(pin.Address-70)), pin.BitPosition)
+	dio, err := pin.ControlChip.findDIODevice(pin.Address)
 	if err != nil {
+		pin.ControlChip.logger.Debug("pin is not from the DIO")
 		return err
 	}
+
+	// if the requested pin is checking the Output WORD. The WORD takes up to 2 bytes
+	pin.outputOffset = dio.i16uOutputOffset
+	pin.inputOffset = dio.i16uInputOffset
+
+	val := false
+
+	// for output gpio pins pwm can be enabled, so we should check for that
+	if pin.isOutputWord() {
+		// if the normal gpio output is given, use the bit position to check if we are in pwm mode
+		val, err = pin.ControlChip.getBitValue(int64(dio.i16uInputOffset+outputPWMActiveOffset+pin.Address%dio.i16uOutputOffset), pin.BitPosition)
+		if err != nil {
+			return err
+		}
+	} else if pin.isOutputPWM() {
+		// we want to read a bit from OutputPWMActive WORD to see if pwm is enabled,
+		// so we convert the pin address into the matching bits, where PWM_1 corresponds to bit 0.
+		// Output pins start at dio.i16uOutputOffset+2, so we can subtract pin address by that amount to get the correct bit
+		pwmActiveBitPosition := uint8(pin.Address - dio.i16uOutputOffset - outputWordToPWMOffset)
+		val, err = pin.ControlChip.getBitValue(int64(dio.i16uInputOffset+outputPWMActiveOffset+uint16(pwmActiveBitPosition/8)),
+			pwmActiveBitPosition%8)
+		if err != nil {
+			return err
+		}
+	}
+
 	pin.pwmMode = val
 	pin.initialized = true
-	pin.ControlChip.logger.Infof("Pin initialized: %#v", pin)
+
+	pin.ControlChip.logger.Debugf("Pin initialized: %#v", pin)
 	return nil
 }
 
 // Get the memory address to use for modifying the PWM duty cycle.
 func (pin *gpioPin) getPwmAddress() uint16 {
-	// The address for the pin is either 70 or 71,
-	// the bit position is used to figure out the offset from the base address of 72 for PWM
-	return pin.Address + 2 + (uint16(pin.BitPosition) % 8) + ((pin.Address - 70) * 7)
+	// The address for the pin is either 0 or 1 + the offset,
+	// the bit position is used to figure out the offset from the base address of 2 + offset for PWM
+	return pin.Address + 7*pin.Address%pin.outputOffset + outputWordToPWMOffset + (uint16(pin.BitPosition))
 }
 
 // Get the memory address to use for modifying the pin state (on/off).
 func (pin *gpioPin) getGpioAddress() uint16 {
-	return pin.Address + (uint16(pin.BitPosition) / 8)
+	switch {
+	case pin.isOutputPWM():
+		return pin.outputOffset + (pin.Address-outputWordToPWMOffset)%pin.outputOffset/8
+	case pin.isInputCounter():
+		// shift the input counter to be a value from 0 to 63, then divide by 32 to get the 0 or 1 address
+		return pin.inputOffset + (pin.Address-pin.inputOffset-inputWordToCounterOffset)/32
```

Comment:
These seem magical to me. Consider putting a comment somewhere explaining where they came from (or at least a comment saying where the documentation that explains them is).

None of the new cases, including the default case, match the original code this switch statement replaced. Is that expected? (i.e., was the previous code wrong in all cases?)

Reply:
spoke offline on this, leaving the links in the readme